examples/fastapi-embeddings-rag/app/main.py

```diff
@@ -0,0 +1,389 @@
+    TABLE.add(items)
+    return {"status": "success"}
+
+
+# ## Retrieval augmented generation (RAG) steps:
+# Now that we've defined our services and created an endpoint to populate data for retrieval, the remaining
+# components of the application will focus on the generative phases of the RAG app.
+
+# ### Retrieval with Sentence Transformers and LanceDB
+# In the retrieval phase we'll first use the Embedder service to create an embedding from the input text
+# to search our LanceDB vector database with. LanceDB is optimized for vector searches in this manner.
+async def retrieve_documents(text: str, limit: int) -> List[Item]:
+    """Retrieve documents from vector DB related to input text"""
+    try:
+        # Encode the input text into a vector
+        vector = await EMBEDDER.encode_text(
+            text, normalize_embeddings=True, run_async=True, stream_logs=False
+        )
+        # Search LanceDB for nearest neighbors to the vector embed
+        results = TABLE.search(vector).limit(limit).to_pydantic(Item)
+        return results
+
+    except Exception as e:
+        raise HTTPException(
+            status_code=500, detail=f"Failed to retrieve documents: {str(e)}"
+        )
+
+
+# ### Format an Augmented Prompt
+# To leverage the documents retrieved from the previous step, we'll format a prompt that provides text from
+# related documents as "context" for the LLM prompt. This allows a general purpose LLM (like Llama) to provide
+# more specific responses to a particular question.
+async def format_prompt(text: str, docs: List[Item]) -> str:
+    """Retrieve documents from vector DB related to input text"""
+    context = "\n".join([doc.page_content for doc in docs])
+    prompt = PROMPT_TEMPLATE.format(question=text, context=context)
+    return prompt
+
+
+# ### Generation GET endpoint
+# Using the methods above, this endpoint will run inference on our LLM to generate a response to a question.
+# The results are enhanced by first retrieving related documents from the source URLs fed into the POST endpoint.
+# Content from the fetched documents is then formatted into the text prompt sent to our self-hosted LLM.
+# We'll be using a generic prompt template to illustrate how many "chat" tools work behind the scenes.
+@app.get("/generate")
+async def generate_response(text: str, limit: int = 4):
+    """Generate a response to a question using an LLM with context from our database"""
+    if not text:
+        return {"error": "Question text is missing"}
+
+    try:
+        # Retrieve related documents from vector DB
+        documents = await retrieve_documents(text, limit)
+
+        # List of sources from retrieved documents
+        sources = set([doc.url for doc in documents])
+
+        # Create a prompt using the documents and search text
+        prompt = await format_prompt(text, documents)
+
+        response = await LLM.generate(
+            prompt=prompt, temperature=0.8, top_p=0.95, max_tokens=100
```

Comment:
maybe we move these other args into the function signature and add a **kwargs to support some of the other params the user can play with?

Reply:
Since these are all optional, I didn't want to explicitly put them in the function signature, but I did include a link to all the available params on vllm to show all available options